### PR TITLE
Add UTMOSFilterStage for audio quality filtering

### PR DESCRIPTION
**Pull request by shubhamNvidia**
## Summary

- Add `UTMOSFilterStage` for filtering audio segments based on UTMOS (UTokyo-SaruLab) predicted Mean Opinion Score (MOS). Uses the `utmos22_strong` model via `torch.hub` from `tarepan/SpeechMOS:v1.2.0`, with automatic resampling to 16 kHz and GPU/CPU support via resource configuration.
- Add `UTMOSConfig` dataclass with configurable `mos_threshold` (default 3.5).
- Add shared audio utilities in `common.py` (`load_audio_file`, `ensure_waveform_2d`, `ensure_mono`, `resolve_model_path`) to reduce code duplication across audio stages.
- Add unit tests for `UTMOSFilterStage` covering pass/fail thresholds, disabled thresholds, multiple items, empty batches, prediction errors, missing waveform fallback, and model unavailability.

## Key features

- Accepts in-memory `waveform + sample_rate` or `audio_filepath` input
- Handles multi-channel audio (converts to mono internally)
- Configurable MOS threshold (set to `None` to pass all)

## Test plan

- [x] `test_process_good_score_passes` -- MOS above threshold passes
- [x] `test_process_bad_score_filtered` -- MOS below threshold filtered
- [x] `test_process_none_threshold_passes_all` -- disabled threshold passes all
- [x] `test_process_multiple_items` -- batch with mixed scores
- [x] `test_process_empty_batch` -- empty batch returns empty
- [x] `test_process_prediction_error_skipped` -- model errors handled gracefully
- [x] `test_no_waveform_no_filepath_skipped` -- missing audio skipped with warning
- [x] `test_model_not_available` -- stage handles unavailable model

### CODE REVIEW

File: nemo_curator/stages/audio/common.py [hunk 1]
```diff
@@ -118,3 +121,95 @@ def process_dataset_entry(self, data_entry: AudioBatch) -> list[AudioBatch]:
...
+    """Resolve a relative model path using the reference file's directory and module subdirectory."""
+    if os.path.isabs(model_path):
+        return model_path
+    current_dir = os.path.dirname(os.path.abspath(reference_file))
+    module_dir = os.path.join(current_dir, module_subdir)
+    for base in (module_dir, current_dir):
+        resolved = os.path.join(base, model_path)
+        if os.path.exists(resolved):
+            return resolved
+    return os.path.join(module_dir, model_path)
```
[the remark this answers is not in the document]
Comment (shubhamNvidia) [reply nemo_curator/stages/audio/common.py:200]: This code is used by other stages in different PR, this code kept same for all the PR

File: tests/stages/audio/filtering/test_utmos.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        stage._model = model
+
+        batch = AudioBatch(
+            data=[_make_item(), _make_item(), _make_item()],
+            task_id="test",
+            dataset_name="test",
+        )
+        result = stage.process(batch)
+
+        assert len(result.data) == 2
```
[the remark this answers is not in the document]
Comment (shubhamNvidia) [reply tests/stages/audio/filtering/test_utmos.py]: Done

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+        try:
+            if sr != self.sample_rate:
+                waveform = torchaudio.transforms.Resample(sr, self.sample_rate)(waveform)
+
+            device = next(self._model.parameters()).device
+            waveform = waveform.to(device)
+
+            with torch.no_grad():
+                score = self._model(waveform, sr=self.sample_rate)
```
[the remark this answers is not in the document]
Comment (shubhamNvidia) [reply nemo_curator/stages/audio/filtering/utmos.py:229]: Not needed

File: tests/stages/audio/filtering/test_utmos.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+    def test_stage_outputs(self) -> None:
+        stage = UTMOSFilterStage()
+        _, outputs = stage.outputs()
+        assert "utmos_mos" in outputs
+
+    def test_stage_inputs(self) -> None:
+        stage = UTMOSFilterStage()
+        required, _ = stage.inputs()
+        assert "data" in required
```
Comment (sarahyurick) [inline tests/stages/audio/filtering/test_utmos.py]: These can be removed, thanks.

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        if self.config is not None:
+            self.mos_threshold = self.config.mos_threshold
+
+    def inputs(self) -> Tuple[List[str], List[str]]:
+        return ["data"], []
+
+    def outputs(self) -> Tuple[List[str], List[str]]:
+        return [], ["utmos_mos"]
+
+    def setup(self, worker_metadata=None) -> None:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]: `setup_on_node` can also be added to download the model.

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+    def setup_on_node(self, _node_info: NodeInfo | None = None, _worker_metadata: WorkerMetadata | None = None) -> None:  # noqa: ARG002
+        """Download the UTMOS model repo via torch.hub (once per node)."""
+        try:
+            torch.hub.load(
+                _UTMOS_REPO, _UTMOS_ENTRYPOINT,
+                trust_repo=True, force_reload=False, skip_validation=True,
+            )
+        except Exception:
+            logger.warning("UTMOS repo pre-download in setup_on_node failed; will retry in setup().")
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]: IMO we should not retry in `setup`. Setup should only work if the model is already downloaded.

File: tests/stages/audio/filtering/test_utmos.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        stage.teardown()
+        assert stage._model is None
+
+    def test_default_threshold(self) -> None:
+        stage = UTMOSFilterStage()
+        assert stage.mos_threshold == 3.5
+
+    def test_custom_threshold(self) -> None:
+        stage = UTMOSFilterStage(mos_threshold=2.0)
+        assert stage.mos_threshold == 2.0
```
Comment (sarahyurick) [inline tests/stages/audio/filtering/test_utmos.py]: These can be removed.

File: nemo_curator/stages/audio/filtering/utmos.py (hunk 6 above)
Comment (shubhamNvidia) [reply nemo_curator/stages/audio/filtering/utmos.py]: removed

**Comment by shubhamNvidia** [comment]
@sarahyurick I’ve merged with main and resolved the conflicts—you can go ahead and run the Ruff linter now

**Comment by sarahyurick** [comment]
Hi @shubhamNvidia the Ruff list is here: https://github.com/NVIDIA-NeMo/Curator/actions/runs/23552052651/job/68568456472?pr=1639

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        self._model_failed = False
+        self._resamplers: dict[int, torchaudio.transforms.Resample] = {}
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return [], []
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        return [], ["utmos_mos"]
+
+    def setup_on_node(self, _node_info: NodeInfo | None = None, _worker_metadata: WorkerMetadata | None = None) -> None:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]:
```suggestion
    def setup_on_node(self, _: NodeInfo | None = None, _: WorkerMetadata | None = None) -> None:
```

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        try:
+            torch.hub.load(
+                _UTMOS_REPO, _UTMOS_ENTRYPOINT,
+                trust_repo=True, force_reload=False, skip_validation=True,
+            )
+        except Exception:  # noqa: BLE001
+            logger.warning("UTMOS repo pre-download in setup_on_node failed.")
+
+    def setup(self, _worker_metadata: WorkerMetadata | None = None) -> None:
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]:
```suggestion
    def setup(self, _: WorkerMetadata | None = None) -> None:
```

**Comment by sarahyurick** [comment]
Hi @shubhamNvidia please update: https://github.com/NVIDIA-NeMo/Curator/actions/runs/23809803316/job/69393484748?pr=1639

File: nemo_curator/stages/audio/filtering/utmos.py (hunk 8 above)
Comment (shubhamNvidia) [reply nemo_curator/stages/audio/filtering/utmos.py]: Python doesn't allow duplicate _ parameter names, it raises SyntaxError: duplicate argument '_' in function definition

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+import torch
+from loguru import logger
+
+try:
+    import torchaudio
+
+    _TORCHAUDIO_AVAILABLE = True
+except ImportError:
+    torchaudio = None
+    _TORCHAUDIO_AVAILABLE = False
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]: I think this can be removed once https://github.com/NVIDIA-NeMo/Curator/pull/1574 is merged.

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+            return None
+
+        try:
+            device = next(self._model.parameters()).device
+            waveform = waveform.to(device)
+
+            if sr != self.sample_rate:
+                if not _TORCHAUDIO_AVAILABLE:
+                    msg = "torchaudio is required for resampling. Install it with: pip install torchaudio"
+                    raise ImportError(msg)  # noqa: TRY301
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]: Same comment as above.

File: nemo_curator/stages/audio/filtering/utmos.py (hunk 10 above)
Comment (shubhamNvidia) [reply nemo_curator/stages/audio/filtering/utmos.py]: noted

File: nemo_curator/stages/audio/filtering/utmos.py (hunk 11 above)
Comment (shubhamNvidia) [reply nemo_curator/stages/audio/filtering/utmos.py]: noted

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+from loguru import logger
+
+try:
+    import torchaudio
+
+    _TORCHAUDIO_AVAILABLE = True
+except ImportError:
+    torchaudio = None
+    _TORCHAUDIO_AVAILABLE = False
+
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]:
```suggestion
import numpy as np
import torch
import torchaudio
from loguru import logger
```

File: nemo_curator/stages/audio/filtering/utmos.py (hunk 11 above)
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]:
```suggestion
```

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+        return self._process_single(task) or []
+
+    def _process_single(self, task: AudioTask) -> AudioTask | None:
+        """Run UTMOS scoring on a single (non-nested) task."""
+        audio_result = _load_waveform_tensor(task.data, task.task_id)
+        if audio_result is None:
+            return None
+        waveform, sr = audio_result
+
+        self._ensure_model()
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py]:
```suggestion
```

File: nemo_curator/stages/audio/filtering/utmos.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+
+import os
+from dataclasses import dataclass, field
+from typing import Any
+
+import numpy as np
+import torch
+import torchaudio
+from loguru import logger
+from nemo_curator.backends.base import NodeInfo, WorkerMetadata
```
Comment (sarahyurick) [inline nemo_curator/stages/audio/filtering/utmos.py:45]:
```suggestion
from loguru import logger

from nemo_curator.backends.base import NodeInfo, WorkerMetadata
```
